postgres: implemented new URL design

Zyqsempai opened the pull request:
Fixes #1175 

Implemented new [URL design](https://github.com/google/go-cloud/blob/master/internal/docs/design.md#urls) for postgres package.
In RDS postgres package implemented all necessary methods for using new URL design, added new temporary method `Open With Url` to not break backward compatibility at this step.

Open questions:
How to handle `traceOpts` which are pathed by Params struct?

Zyqsempai commented:
cc @zombiezen

zombiezen commented:
(I'm currently reviewing this PR. I plan to send my comments back on Monday.)

zombiezen commented on postgres/rdspostgres/rdspostgres.go:
```diff
@@ -53,6 +55,41 @@ type Params struct {
+}
+
+// OpenWithUrl temporary Open DB method, for smooth migration to new URL based design.
+func OpenWithUrl(provider rds.CertPoolProvider, url *url.URL) (*sql.DB, error) {
```
I don't understand why this function is exported; can you help me understand what the purpose of it is?

Zyqsempai replied:
Oh, yeah, you are right, just сarelessness from my side i suppose;)

zombiezen commented on postgres/postgres.go:
```diff
@@ -0,0 +1,96 @@
+	return opener.OpenConnectionURL(ctx, u)
+}
+
+var defaultURLMux = new(URLMux)
```
I would also expect the default URL mux to include a `postgres://` opener. In fact, I would recommend removing the RDS changes from this initial PR and focusing on getting a URL opener for plain Postgres first. We can add RDS and Cloud SQL as follow-up PRs, and they should go pretty smoothly since we would already figure out most of the details with smaller scale.

Zyqsempai replied:
Ok, would you suggest to implement `plain Postgres` Opener right in postgres package? Or create the third one?

zombiezen replied:
Right in the postgres package seems correct to me.

Zyqsempai commented:
Implemented simpliest plain Postgres UrlOpen.
Added probably temporary tests for local Postgres test connection.
Fixed other PR comments.

Open questions:
Should we validate url somehow?

zombiezen commented on postgres/postgres.go:
```diff
@@ -0,0 +1,96 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package postgres
```
No blank line between the package comment and the package declaration.

zombiezen commented on postgres/postgres.go:
```diff
@@ -0,0 +1,125 @@
+
+// OpenWithUrl temporary Open DB method, for smooth migration to new URL based design.
+func openWithUrl(url *url.URL) (*sql.DB, error) {
+	return sql.Open(Scheme, url.String())
```
This should use [`ocsql.Wrap`](https://godoc.org/contrib.go.opencensus.io/integrations/ocsql#Wrap) to add the automatic OpenCensus instrumentation.

zombiezen commented:
> Open questions:
> Should we validate url somehow?

I think it's fine to pass the URL through and let the PostgreSQL driver take care of validating the connection string.

Zyqsempai commented:
@zombiezen Done

zombiezen commented on postgres/postgres.go:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	// Wrap the driver.Connector with ocsql.
+	connector = ocsql.WrapConnector(connector, ocsql.WithAllTraceOptions())
```
This function isn't in a released version of `ocsql` yet, so tests are failing.

Zyqsempai commented:
@zombiezen Switched to the old/good `Wrap()` function, sorry that I didn't listen to you from the first time;)
